src/pmsg.rs

```diff
@@ -0,0 +1,163 @@
+#[cfg(target_os = "android")]
+pub mod android {
+    use crate::{logging_iterator::NewlineScaledChunkIterator, thread, Buffer, Priority};
+    use bytes::{BufMut, BytesMut};
+    use std::{
+        fs::{File, OpenOptions},
+        io::{self, Write},
+        path::Path,
+        time::SystemTime,
+    };
+
+    /// Persistent message charater device
+    const PMSG0: &str = "/dev/pmsg0";
+
+    /// 'Magic' marker value of android logger
+    const ANDROID_LOG_MAGIC_CHAR: u8 = b'l';
+    /// Maximum size of log entry payload
+    const ANDROID_LOG_ENTRY_MAX_PAYLOAD: usize = 4068;
+    /// Increment of sequence number when breaking messages in the Android logging system
+    const ANDROID_LOG_PMSG_SEQUENCE_INCREMENT: usize = 1000;
+    // Maximum sequence number in Android logging system
+    const ANDROID_LOG_PMSG_MAX_SEQUENCE: usize = 256000;
+
+    lazy_static::lazy_static! {
+        static ref PMSG_DEV: PmsgDev = PmsgDev::connect(Path::new(PMSG0));
+    }
+
+    /// Persistent message character device abstraction. Can only be written to
+    /// and not read from.
+    struct PmsgDev {
+        file: parking_lot::RwLock<File>,
+    }
+
+    impl PmsgDev {
+        /// Construct a new PmsgDev.
+        ///
+        /// # Panics
+        ///
+        /// This function panics if the underlying character device cannot be opened
+        /// for writing.
+        pub fn connect(path: &Path) -> Self {
+            let pmsg_dev = OpenOptions::new().write(true).open(path).expect("failed to open pmsg device");
+
+            Self {
+                file: parking_lot::RwLock::new(pmsg_dev),
+            }
+        }
+
+        /// Write a buffer to the pmsg device.
+        ///
+        /// Structure and length limits for meaningful input has to be handled by
+        /// the calling side.
+        pub fn write_all(&self, buffer: &[u8]) -> io::Result<()> {
+            let mut pmsg = self.file.write();
+            pmsg.write_all(buffer)
+        }
+
+        /// Flush the backing file handle.
+        pub fn flush(&self) -> io::Result<()> {
+            let mut pmsg = self.file.write();
+            pmsg.flush()
+        }
+    }
+
+    /// Send a log message to pmsg0
+    pub(crate) fn log(tag: &str, buffer_id: Buffer, priority: Priority, message: &str) {
+        // TODO: The timestamps of logd and pmsg won't match if we create them separately
+        let timestamp_as_duration = SystemTime::now()
+            .duration_since(std::time::UNIX_EPOCH)
+            .expect("failed to aquire time");
+        let timestamp_secs = timestamp_as_duration.as_secs() as u32;
+
+        // TODO: Fetch real uid
+        let uid = 0;
+        let pid = std::process::id() as u16;
+        let thread_id = thread::id() as u16;
```

Comment:
Create Record structure for `pid`, `threadId`, `timestamp`, `tag`, `priority`, `buffer`, `message`

Reply:
Order of output: timestamp, pid, thread_id, buffer, tag, priority, message